### PR TITLE
[DNM] raftstore, storage: check key range when region epoch is stale.

### PR DESCRIPTION
**Pull request by disksing**
This PR aims to reduce data retransmission when region splits.

@siddontang @zhangjinpeng1987 @BusyJay

### CODE REVIEW

File: src/raftstore/store/store.rs [hunk 1]
```diff
@@ -994,19 +999,24 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
 
-        let res = peer.check_epoch(msg);
-        if let Err(Error::StaleEpoch(msg, mut new_regions)) = res {
+        let res = peer.check_epoch(msg, updated_regions);
+        if let Err(Error::StaleEpoch(msg)) = res {
             // Attach the next region which might be split from the current region. But it doesn't
             // matter if the next region is not split from the current region. If the region meta
             // received by the TiKV driver is newer than the meta cached in the driver, the meta is
             // updated.
             if let Some((_, &next_region_id)) = self.region_ranges
```
Comment (ngaut) [inline src/raftstore/store/store.rs]: Do we still need next region id. We do have the stale epoch region id.

Comment (disksing) [inline src/raftstore/store/store.rs]: Yes. If not, client will have to request next region from PD server.

File: src/raftstore/store/peer.rs [hunk 2]
```diff
@@ -923,12 +929,27 @@ impl Peer {
...
+            let mut updated_region = UpdatedRegion::new();
+            updated_region.set_region(latest_region.clone());
+            updated_region.set_leader(self.peer.clone());
+            updated_regions.push(updated_region);
+
+            if req.get_header().has_key_range() {
+                let range = req.get_header().get_key_range();
+                if range.get_min_key() >= latest_region.get_start_key() &&
+                   range.get_max_key() < latest_region.get_end_key() {
+                    debug!("{} accept stale epoch {:?} after check key range",
```
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: s/check/checking/

File: src/raftstore/store/peer.rs [hunk 3]
```diff
@@ -1343,9 +1369,10 @@ impl<'a> ExecContext<'a> {
 impl Peer {
     // Only errors that will also occur on all other stores should be returned.
     fn exec_raft_cmd(&mut self,
-                     ctx: &mut ExecContext)
+                     ctx: &mut ExecContext,
+                     updated_regions: &mut Vec<UpdatedRegion>)
```
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: Why `Vec`?

File: src/raftstore/store/peer.rs [hunk 4]
```diff
@@ -691,13 +691,16 @@ impl Peer {
 
             // Execute ready-only commands immediately in leader when doing local reads.
             let mut ctx = ExecContext::new(self, 0, 0, &req);
-            let (mut resp, _) = self.exec_raft_cmd(&mut ctx).unwrap_or_else(|e| {
-                error!("{} execute raft command err: {:?}", self.tag, e);
-                (cmd_resp::new_error(e), None)
-            });
+            let mut updated_regions = vec![];
```
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: I think this can be put into `ExecContext`.

**Comment by BusyJay** [comment]
You may need to explain more in your pr description.

**Pull request closed without merging**